### PR TITLE
DS4 support and XInput rumble

### PR DESCRIPTION
**Pull request by jarveson**
Adds native ds4 input support. Sixaxis / rumble seems to work fairly well with various games. It also should work with both usb and/or bluetooth connection. Thanks for @Zangetsu38 for most of the testing

If we want the hidapi an actual submodule I can change it, but i added in cmake support and updated the vs project in it so it can't be off of the actual repo.

On paper this should also work on linux, but its untested. Also results may vary with ds4windows running as    we seem to fight it a bit for rumble/light color

### CODE REVIEW

**Comment by danilaml** [comment]
@jarveson wouldn't it be better to ad hidapi fork to RPCS3 org as we did with some other used projects, create a branch with our custom changes there and use this ad a submodule? Also, preferably submit the PR with our updates to the upstream.

**Comment by Zangetsu38** [comment]
@danilaml Offficial git hidapi it is this https://github.com/signal11/hidapi ?

**Comment by Nekotekina** [comment]
It should be a submodule, otherwise it's very hard to tell what is modified and harder to update.
Why cmake/vs files cannot be outside? In this case we don't even need to fork.

**Comment by deadair3210** [comment]
Would native DS3 support be too far off of a request? Would be nice to be able to use the DS3 for the PS3

**Comment by mirh** [comment]
DS3 isn't natively exposed to the OS in the first place [except perhaps raw input? Dunno(and if we are just talking about windows ofc)]. 
You require a custom mini-hid driver afaik, at which point I'm not sure XInput/DirectInput support should be all that different from other controllers.

**Comment by jarveson** [comment]
^ Yea, i don't have a ds3 and the protocol is different enough that it would have to be looked into separately

Regarding the submodule, the hidapi repo itself is fairly stable and doesnt look to be updated that often so i doubt there will be any actual major changes to deal with. You're right in that the cmake / vs files could be pulled out but i think having the whole nested folders / double folder to deal with it is worse and just adds clutter

Anyway, i've forked the repo and committed the changes needed to hidapi, https://github.com/jarveson/hidapi , if you want to fork/clone this under rpcs3 team repos ill make the appropriate changes to this PR, or however you want to do it

**Comment by deadair3210** [comment]
Forgot that ds3 wasn't exposed natively to windows, I forgot I had installed drivers to get it working

⁣Sent from BlueMail ​

**Comment by danilaml** [comment]
@jarveson The CMakeLists from here seems to be much more complete/robust: https://github.com/signal11/hidapi/pull/33/files
Also, would be gut to also add new lib dependencies (for OSX and linux) in the readme.
> having the whole nested folders / double folder to deal with it is worse and just adds clutter

In what way? It's a bit less elegant, yes. But certainly better, than just plain pasting the whole folder (just look at the PR diff...). Though just using the fork for the submodule seems to be the cleanest solution and easiest to deal with, provided we don't need to update it often.

**Comment by mirh** [comment]
> ds4: fix bluetooth vs wired detection

I'm not sure that's meant to be applications business?

**Comment by jarveson** [comment]
It is if you want bluetooth connections to work

**Comment by mirh** [comment]
I meant: I'm relatively sure the [driver](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d2d782fccee4f699a35e2d0cdbb2b19bdaec95a4) should be [logically] supposed to handle this kind of problem.

**Comment by jarveson** [comment]
This operates below hid drivers

File: rpcs3/DS4PadHandler.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		std::string path;
+		bool btCon;
+		u8 largeVibrate;
+		u8 smallVibrate;
+	};
+
+	const u16 DS4_VID = 0x054C;
+
+	// pid's of connected ds4
+    const std::array<u16, 3> ds4Pids = {{0xBA0, 0x5C4, 0x09CC }};
```
Comment (aclonegeek) [inline rpcs3/DS4PadHandler.h]: Looks like our friend here could use a few more spaces.

Comment (danilaml) [inline rpcs3/DS4PadHandler.h]: tabs (be they damned)

**Comment by hcorion** [comment]
Could this be extended into support for general controllers under Linux, or would it be better to continue with the native evdev implementation?

**Comment by jarveson** [comment]
It would definitely be better to go with native evdev still, this commit just operates on raw packets and ignores any of the specs that the device gives, so its very specific to ds4, which is the only nice way to make it work on windows natively.  Linux gets around this by patching the hid descriptor for sony devices in a kernel driver, so evdev should work just as well as this (if it were implemented)

**Pull request merged**